### ISSUE TITLE
Installation on Docker fails 

### ISSUE THREAD

**Issue by juandelgado** [rvm/ubuntu_rvm#8]
Hello,

I'm trying to get RVM installed on Ubuntu, but using Docker which seems to complicate things quite a bit. This is the smallest Dockerfile that can reproduce the problem:

```
FROM ubuntu:16.04
RUN apt-get update && \
    apt-get install -y software-properties-common
RUN apt-add-repository -y ppa:rael-gc/rvm && apt-get update
RUN apt-get install -y rvm
```
You can build that Dockerfile with something like this (Docker must be already installed):

`docker build -t ubuntu-rvm .`

`apt-get install -y rvm` fails. This is the problem, I think:

```bash
/var/lib/dpkg/info/rvm.postinst: line 25: /etc/profile.d/rvm.sh: No such file or directory
dpkg: error processing package rvm (--configure):
 subprocess installed post-installation script returned error exit status 1
```

Anything obvious that pops?

A few things to note: 
 * `apt-add-repository` seems not to be part of a vanilla Ubuntu 16.04 installation, that's why `software-properties-common` is installed first. I was thinking about sending a PR to update the docs to mention this?
 * The above Docker file will still not leave a running RVM installation because it's missing the "run command as login shell bit", which I think it can be done in the terminal with `bash -l`, but I'm stuck on the problem above.

**Comment by raelgc** [comment]
Hi @juandelgado and thanks for raising this.

Indeed, `apt-add-repository` was not part of the first versions and it was included later. I like the idea about create a PR to mention it in docs! Please, go ahead!

The error is saying `/etc/profile.d/rvm.sh` is not present, but it should be created by `rvm` install (which is called before the `postinst` script.

Has your install failed before this error?

**Comment by juandelgado** [comment]
Hey hey.

You can see the full log in [this gist](https://gist.github.com/juandelgado/a26f05acb0bc4ba80847110b41829edc). There's nothing particularly wrong, other than some errors trying to configure dialogs, because this is all being run from the command line and inside Docker.

I thought the file would be created or copied before, but I [searched the repo for "rvm.sh"](https://github.com/rvm/ubuntu_rvm/search?utf8=%E2%9C%93&q=%22rvm.sh%22&type=) and I'm not quite sure when that happens. I might be missing something!

I'll prepare that PR and send it over.

Thanks!

Juan

**Comment by raelgc** [comment]
This repo is only for the Ubuntu package, so I don't include the RVM files to avoid duplication.

Anyway, I just updated Xenial package to use rvm `1.29.2`.

I can see [in the installer](https://github.com/rvm/rvm/blob/8cba2caf709a13283ec534a38793e4aa27d7fab1/scripts/functions/installer) that `rvm` will use user home profile if `rvm_path` is pointing to it.

Is there any setup for `rvm_path` or something blocking the user to install `rvm` globally?

**Comment by juandelgado** [comment]
> Is there any setup for rvm_path or something blocking the user to install rvm globally?

I'm not sure I understand the question. 

If you are asking for my personal use case, then no. I'd be happy with whatever as long as it is a working RVM setup. The only thing to consider is that Docker is logged in as root by default which indeed seems problematic as per [RVM's docs](https://rvm.io/support/troubleshooting#sudo) (mentions our missing file):

> do not run the installer directly as/from the 'root' account! Always use the sudo command from a non-privileged user account. This is due to variables that are different between root's environment and a user's that aren't affected by an EUID change, as well as code checks in the install itself. In addition, using sudo will not load the /etc/profile.d/rvm.sh, which exists for current installs, and as such will not load RVM.

Not sure where to go now : /

**Comment by raelgc** [comment]
Maybe I should remove the call to `source /etc/profile.d/rvm.sh` from the `postinst` script.

**Comment by raelgc** [comment]
@mpapis, any thoughts? :)

**Comment by mtarvainen** [comment]
Hi there!

I've ran into same issue as described here. As a quick&dirty way to get forward I created `/etc/profile.d/rvm.sh` file manually and got following notice:
```
  * WARNING: your RVM loading script `/etc/profile.d/rvm.sh` is deprecated
    and causes you to have `umask g+w` set in your shell,
    run `rvm get head --auto-dotfiles` again to fix your scripts.
```
I think it would be reasonable to remove that call from `postinst` script or add if statement that checks whether the file is present before calling it.

**Comment by raelgc** [comment]
My proposed solution to the team (waiting feedback from @mpapis or @pkuczynski):

1. remove this `source` call, and
2. replace the info note in the package post install (to make clear user needs to logout/login)
2. a minor: remove the message about rvm no more adding user to groups, as the package does this

**Comment by stepanselyuk** [comment]
In a Dockerfile (based on Ubuntu 16.04) I put it as:

```
RUN apt-get install -y software-properties-common \
    && apt-add-repository -y ppa:rael-gc/rvm \
    && apt-get update \
    # hack for https://github.com/rvm/ubuntu_rvm/issues/8
    && touch /etc/profile.d/rvm.sh \
    && apt-get install -y rvm \
    && sudo -E bash -c 'source /usr/share/rvm/scripts/rvm && rvm install ruby && gem install compass bundler --no-ri --no-rdoc'
```
Though not sure if `sudo -E` is needed.

**Comment by ggeldenhuis** [comment]
The same issue exists when installing rvm on Ubuntu 17.04. I am using puppet to install the ppa and then puppet again to install the rvm package. My puppet run fails because deb reported a problem
`/var/lib/dpkg/info/rvm.postinst: line 25: /etc/profile.d/rvm.sh: No such file or directory
dpkg: error processing package rvm (--configure):
 subprocess installed post-installation script returned error exit status 1
Errors were encountered while processing:
 rvm
E: Sub-process /usr/bin/dpkg returned an error code (1)`

**Comment by raelgc** [comment]
@ggeldenhuis For now, a workaround is to add a step to `touch /etc/profile.d/rvm.sh` like suggested by @stepanselyuk.

**Comment by ggeldenhuis** [comment]
@raelgc I tried that but I then just get other errors. I have not investigated it much further though. Will try and spend some more time on it tomorrow and post any other errors.

**Comment by raelgc** [comment]
The yesterday RVM release is already using the package with the fix.

**Comment by juandelgado** [comment]
Hey hey, thanks for the work on this! 

Gave it a go, the original Docker file posted in the issue no longer errors out. This is the end of the installation log now:

```
Installing RVM to /usr/share/rvm/
    Adding rvm PATH line to /root/.profile /root/.mkshrc /root/.bashrc /root/.zshrc.
    Adding rvm loading line to /root/.profile /root/.bash_profile /root/.zlogin.
Installation of RVM in /usr/share/rvm/ is almost complete:
  * To start using RVM you need to run `source /usr/share/rvm/scripts/rvm`
    in all your open shell windows, in rare cases you need to reopen all shell windows.
```

Unfortunately `source` is not available on typical Docker installs and trying something like:

```bash
/bin/bash -c "source /usr/share/rvm/scripts/rvm"
rvm install 2.3
```

Throws an error: `rvm: not found`

It seems that `rvm` is just not in the `$PATH`. Calling `source` is not helping.

Thoughts?

Thanks!

Juan

**Comment by raelgc** [comment]
Usually, for a normal user, a logout/login is required.

**Comment by juandelgado** [comment]
Unfortunately @raelgc that's typically not an option when you are running inside Docker.

Is there anything that comes to mind that you think we could try?

Thanks,

**Comment by raelgc** [comment]
@juandelgado I think this is out of the scope of this issue, because you need to relogin to be part of the `rvm` group and read the scripts. But there are [some alternatives](https://superuser.com/questions/272061/reload-a-linux-users-group-assignments-without-logging-out), like run something like `exec su -l $USER`.

**Comment by raelgc** [comment]
Re-open if this is still an issue.

**Comment by ColinHebert** [comment]
Sorry I might need to reopen that now. We still don't know why `/etc/profile.d/rvm.sh` doesn't get created in the first place, correct?

Some scripts/systems are relying on it.

**Comment by raelgc** [comment]
It's not created because you're running the install as `root`. I've fixed this and you need to use the above linked alternatives to the required logout/login.

**Comment by ColinHebert** [comment]
Sorry I have to ask you to indulge me a few questions.

Why is being root a problem in the first place, the install script handles that as far as I am aware.

Is the solution to do the logout/login before doing `apt-get install`? Can't/shouldn't that be part of the install process itself? It's meant to be installed through `apt`, so the expectation is that this should be executed by root in any case.
Am I missing something?

**Comment by raelgc** [comment]
The install is ran by `root`, but later `rvm` is expecting you a normal user. Anyway, I fixed this few releases ago.

The logout/login is required after the install because you need to be part of the `rvm` group, and joining a new group takes effect only after a relogin.

Anyway, you can try the workarounds I [listed above](https://superuser.com/questions/272061/reload-a-linux-users-group-assignments-without-logging-out) to try reload groups info without logout.

**Comment by raelgc** [comment]
@olegburca check this workaround to make sure your user is listed in the new group: https://superuser.com/questions/272061/reload-a-linux-users-group-assignments-without-logging-out

**Comment by eriko-de** [comment]
Just for anyone who is searching for a way to build a docker image using rvm... Here is my solution, so far:

I was able to install ruby via rvm and using it very seamlessly in a ubuntu docker container by using this Dockerfile:
`FROM ubuntu:16.04

RUN apt-get update && apt-get -y install software-properties-common

##################################### RVM #################################
# Changing shell, needed for rvm to work
SHELL ["/bin/bash", "-cl"]

RUN gpg --keyserver hkp://pool.sks-keyservers.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3 7D2BAF1CF37B13E2069D6956105BD0E739499BDB

RUN apt-add-repository -y ppa:rael-gc/rvm && \
  apt-get update && \
  apt-get -y install rvm
RUN echo 'source "/usr/share/rvm/scripts/rvm"' >> ~/.bashrc

RUN rvm install 2.6.3

# Sanity check
RUN rvm --version && ruby --version
`
The only 'downside' of this is, that you are getting a warning everytime you are using the 'RUN' command:
`mesg: ttyname failed: Inappropriate ioctl for device`

It is realted to the login shell mode of bash, not sure how to prevent this

**Comment by stepanselyuk** [comment]
I checked one of my Dockerfile, and below this is how I solved that issue:
```
RUN apt-get install -y software-properties-common

RUN gpg --keyserver hkp://keys.gnupg.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3 7D2BAF1CF37B13E2069D6956105BD0E739499BDB \
    && curl -sSL https://get.rvm.io | bash -s stable

RUN usermod -a -G rvm root \
    && usermod -a -G rvm youruser

USER youruser

RUN sudo -E bash -c 'source /etc/profile.d/rvm.sh && rvm install ruby 2.7.0 && gem update --system && gem install compass'
```

**Comment by eriko-de** [comment]
@stepanselyuk does this make ruby available in PATH for all subsequent programs and RUN commands?

**Comment by stepanselyuk** [comment]
@eriko-de I did not check this, cause I use absolute paths to ruby and so on binaries. If you need it in PATH you can do like this (again, I did not check this):
```
ENV PATH=$PATH
RUN export PATH="/path/to/add:$PATH"
```

**Comment by biot023** [comment]
This is driving me mad.
Trying to install RVM and 2.5.1 for a user called 'deploy' in a Dockerfile.
Cannot get the group thing working.
Calling `source` on the rvm script doesn't work as 'deploy'.
I'm possibly missing something obvious but I've been going around in circles so long I no longer trust my own judgement!

**Comment by felipelmota** [comment]
@stepanselyuk answer worked for me.
